### PR TITLE
Generate install entries in one pass

### PR DESCRIPTION
**Pull request by rgrinberg**
This is almost ready. I just need to memoize the creation of the map.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Okay, Local_package is now just a simple wrapper to make accessing build paths easier. It can be removed completely but it will introduce a bit of boilerplate I believe.

File: src/install_rules.ml [hunk 1]
```diff
@@ -312,84 +448,16 @@ let promote_install_file (ctx : Context.t) =
...
-  coqlib_install_files
-  |> List.rev_append lib_install_files
-  |> List.rev_append installs
-  |> List.rev_append docs
-  |> List.rev_append metadata
+let install_entries sctx (package : Package.t) =
+  let packages = Stanzas_to_entries.stanzas_to_entries sctx in
+  Package.Name.Map.Multi.find packages package.name
 
 let install_entries =
```
Comment (diml) [inline src/install_rules.ml]: We can remove this. `stanzas_to_entries` is already memoised, so this is caching a `Map.find`, which is likely to be less costly than the overhead of the memoisation system.

File: src/stdune/map.ml [hunk 2]
```diff
@@ -219,6 +219,15 @@ module Make(Key : Key) : S with type key = Key.t = struct
...
     let find t k = Option.value (find t k) ~default:[]
+
+    let add_all t k = function
+      | [] -> t
+      | entries ->
+        update t k ~f:(fun v ->
+          Some (
+            match v with
+            | None -> entries
+            | Some x -> List.rev_append x entries))
```
Comment (diml) [inline src/stdune/map.ml]: We don't know if the order matters here, so we should use `List.append`

**Comment by diml** [comment]
I replaced `Local_package` by a `Package_paths` sub-module in `install_rules.ml` that takes a build context and a `Package.t`. That seems simpler in the end

File: src/install_rules.ml (hunk 1 above)
Comment (diml) [reply src/install_rules.ml]: I ended up doing this and the next change as well

**Comment by rgrinberg** [comment]
Thanks for the fixes

**Pull request merged**